UC4-chaincode/src/main/java/de/upb/cs/uc4/chaincode/contract/admission/AdmissionContractUtil.java

```diff
@@ -79,29 +111,99 @@ public boolean checkStudentMatriculated(ChaincodeStub stub, AbstractAdmission ad
+        } catch (LedgerAccessError e) {
+            return false;
+        }
+    }
+
+    private boolean studentListensToModule(ChaincodeStub stub, String enrollmentId, String moduleId){
         ExaminationRegulationContractUtil erUtil = new ExaminationRegulationContractUtil();
         MatriculationDataContractUtil matUtil = new MatriculationDataContractUtil();
-
         try {
-            MatriculationData matriculationData = matUtil.getState(stub, admission.getEnrollmentId(), MatriculationData.class);
+            MatriculationData matriculationData = matUtil.getState(stub, enrollmentId, MatriculationData.class);
             List<SubjectMatriculation> matriculations = matriculationData.getMatriculationStatus();
             for (SubjectMatriculation matriculation : matriculations) {
                 String examinationRegulationIdentifier = matriculation.getFieldOfStudy();
                 ExaminationRegulation examinationRegulation = erUtil.getState(stub, examinationRegulationIdentifier, ExaminationRegulation.class);
                 List<ExaminationRegulationModule> modules = examinationRegulation.getModules();
                 for (ExaminationRegulationModule module : modules) {
-                    if (module.getId().equals(admission.getModuleId())) {
+                    if (module.getId().equals(moduleId)) {
                         return true;
                     }
                 }
             }
         } catch (LedgerAccessError e) {
             return false;
         }
+        return false;
+    }
+
+    public boolean checkExamAdmittable(ChaincodeStub stub, ExamAdmission admission) {
+        if(!checkExamExists(stub, admission)){
+            return true; // the simpler error is already reported
+        }
+
+        // check
+        try {
+            Exam exam = new ExamContractUtil().getState(stub, admission.getExamId(), Exam.class);
+            return exam.getAdmittableUntil().isAfter(Instant.now());
+        } catch (LedgerAccessError e) {
+            return false;
+        }
+    }
 
+    public boolean checkExamExists(ChaincodeStub stub, ExamAdmission admission) {
+        try {
+            Exam exam = new ExamContractUtil().getState(stub, admission.getExamId(), Exam.class);
+            if (exam != null){
+                return true;
+            }
+        } catch (LedgerAccessError e) {
+            return false;
+        }
         return false;
     }
 
+    public boolean checkExamAdmissionNotAlreadyExists(ChaincodeStub stub, ExamAdmission admission) {
+        if(!checkExamExists(stub, admission)){
+            return true; // the simpler error is already reported
+        }
+
+        // check
+        List<ExamAdmission> examAdmissions = this.getAllStates(stub, ExamAdmission.class);
+        return examAdmissions.stream().noneMatch(item ->
+                item.getExamId().equals(admission.getExamId())
+                && item.getEnrollmentId().equals(admission.getEnrollmentId()));
```

Comment:
We could just build the admission key and query for that directly, but this should of course also work.